src/llama-sampling.cpp

```diff
@@ -1082,6 +1093,64 @@ struct llama_sampler * llama_sampler_init_xtc(float p, float t, size_t min_keep,
     };
 }
 
+// k-shift
+
+struct llama_sampler_k_shift {
+    const int32_t k;
+    bool k_set;
+};
+
+static const char * llama_sampler_k_shift_name(const struct llama_sampler * /*smpl*/) {
+    return "k-shift";
+}
+
+static void llama_sampler_k_shift_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
+    auto * ctx = (llama_sampler_k_shift *) smpl->ctx;
+
+    if (ctx->k_set == true
+        || ctx->k <= 0
+        || ctx->k >= (int) cur_p->size) {
```

Comment:
There appears to be a bug here: If at the first token position, `ctx->k >= (int) cur_p->size` (e.g. because a preceding truncation sampler has already removed too many tokens) then we return and `k_set` remains false. This means that K-Shift will only take effect on the *second* (or later) token of the output, violating its contract.